datacube/scripts/product.py

```diff
@@ -104,30 +112,106 @@ def update_dataset_types(index, allow_unsafe, allow_exclusive_lock, dry_run, fil
+                            'ancillary_quality', 'latgqa_cep90', 'product_type',\
+                            'gqa_abs_iterative_mean_xy', 'gqa_ref_source', 'sat_path',\
+                            'gqa_iterative_stddev_xy', 'time', 'sat_row', 'orbit', 'gqa',\
+                            'instrument', 'gqa_abs_xy', 'crs', 'resolution', 'tile_size',\
+                            'spatial_dimensions'], extrasaction='ignore')
+    writer.writeheader()
+
+    def add_first_name(row):
+        names_ = row['name']
+        row['name'] = names_ if names_ else None
+        return row
+
+    writer.writerows(add_first_name(row) for row in index)
+
+def _write_yaml(index):
     """
-    List products that are defined in the generic index.
+    Dump yaml data with support for OrderedDicts.
+
+    Allows for better human-readability of output: such as dataset ID field first, sources last.
+
+    (Ordered dicts are output identically to normal yaml dicts: their order is purely for readability)
     """
-    products = dc.list_products()
+
+    # We can't control how many ancestors this dumper API uses.
+    # pylint: disable=too-many-ancestors
+    class OrderedDumper(yaml.SafeDumper):
+        pass
+
+    def _dict_representer(dumper, data):
+        return dumper.represent_mapping(yaml.resolver.BaseResolver.DEFAULT_MAPPING_TAG, data.items())
+
+    OrderedDumper.add_representer(OrderedDict, _dict_representer)
+    try:
+        return yaml.dump_all(index, sys.stdout, Dumper=OrderedDumper, default_flow_style=False, indent=4)
+    except TypeError:
+        return yaml.dump(index.definition, sys.stdout, Dumper=OrderedDumper, default_flow_style=False, indent=4)
+
+def _write_tab(products):
+    products = pd.DataFrame(products)
 
     if products.empty:
         echo('No products discovered :(')
         return
 
-    echo(products.to_string(columns=('name', 'description', 'product_type', 'instrument',
-                                     'format', 'platform'),
+    echo(products.to_string(columns=('id', 'name', 'description', 'ancillary_quality',\
+                                     'product_type', 'gqa_abs_iterative_mean_xy',\
+                                     'gqa_ref_source', 'sat_path',\
+                                     'gqa_iterative_stddev_xy', 'time', 'sat_row',\
+                                     'orbit', 'gqa', 'instrument', 'gqa_abs_xy', 'crs',\
+                                     'resolution', 'tile_size', 'spatial_dimensions'),
                             justify='left'))
 
+LIST_OUTPUT_WRITERS = {
+    'csv': _write_csv,
+    'yaml': _write_yaml,
+    'tab': _write_tab,
+}
+
+@product.command('list')
+@click.option('-f', help='Output format',
+              type=click.Choice(list(LIST_OUTPUT_WRITERS)), default='yaml', show_default=True)
+@ui.pass_datacube()
+def list_products(dc, f):
+    """
+    List products that are defined in the generic index.
+    """
+    LIST_OUTPUT_WRITERS[f]((build_product_list(dc.index)
```

Comment:
The double parentheses here are unnecessary and make it look more confusing than needed, this line can be simplified to:
```python
    LIST_OUTPUT_WRITERS[f](build_product_list(dc.index))
```